include/cuco/detail/hyperloglog/hyperloglog_ref.cuh

```diff
@@ -0,0 +1,535 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cuco/cuda_stream_ref.hpp>
+#include <cuco/detail/__config>
+#include <cuco/detail/error.hpp>
+#include <cuco/detail/hyperloglog/finalizer.cuh>
+#include <cuco/detail/hyperloglog/kernels.cuh>
+#include <cuco/detail/utils.hpp>
+#include <cuco/hash_functions.cuh>
+#include <cuco/sketch_size.hpp>
+#include <cuco/utility/cuda_thread_scope.cuh>
+#include <cuco/utility/traits.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/type_traits/is_contiguous_iterator.h>
+
+#include <cuda/std/bit>
+#include <cuda/std/span>
+#include <cuda/std/utility>
+
+#include <cooperative_groups.h>
+#include <cooperative_groups/reduce.h>
+
+#include <cstddef>
+
+namespace cuco::detail {
+
+/**
+ * @brief A GPU-accelerated utility for approximating the number of distinct items in a multiset.
+ *
+ * @note This class implements the HyperLogLog/HyperLogLog++ algorithm:
+ * https://static.googleusercontent.com/media/research.google.com/de//pubs/archive/40671.pdf.
+ *
+ * @tparam T Type of items to count
+ * @tparam Scope The scope in which operations will be performed by individual threads
+ * @tparam Hash Hash function used to hash items
+ */
+template <class T, cuda::thread_scope Scope, class Hash>
+class hyperloglog_ref {
+  // We use `int` here since this is the smallest type that supports native `atomicMax` on GPUs
+  using fp_type = double;  ///< Floating point type used for reduction
+  using hash_value_type =
+    decltype(cuda::std::declval<Hash>()(cuda::std::declval<T>()));  ///< Hash value type
+ public:
+  static constexpr auto thread_scope = Scope;  ///< CUDA thread scope
+
+  using value_type    = T;     ///< Type of items to count
+  using hash_type     = Hash;  ///< Hash function type
+  using register_type = int;   ///< HLL register type
+
+  template <cuda::thread_scope NewScope>
+  using with_scope = hyperloglog_ref<T, NewScope, Hash>;  ///< Ref type with different
+                                                          ///< thread scope
+
+  /**
+   * @brief Constructs a non-owning `hyperloglog_ref` object.
+   *
+   * @throw If sketch size < 0.0625KB or 64B
+   * @throw If sketch storage has insufficient alignment
+   *
+   * @param sketch_span Reference to sketch storage
+   * @param hash The hash function used to hash items
+   */
+  __host__ __device__ constexpr hyperloglog_ref(cuda::std::span<std::byte> sketch_span,
+                                                Hash const& hash)
+    : hash_{hash},
+      precision_{cuda::std::countr_zero(
```

Comment:
The precision value is determined by the size of the `sketch_span` and will select the highest possible precision for the given number of bytes. In other words, the user can adjust the precision by making the span smaller or larger.

I'm planning to add another ctor overload (for the host/bulk API) that takes a maximum relative standard deviation of the error instead of the sketch size. This is actually how Spark implemented [their API](https://github.com/apache/spark/blob/49b4c3bc9c09325de941dfaf41e4fd3a4a4c345f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/HyperLogLogPlusPlusHelper.scala#L29-L42). This change requires a bit more work to validate that we are actually within the bounds of the requested deviation. So this will be a future PR.

Reply:
> 'm planning to add another ctor overload (for the host/bulk API) that takes a maximum relative standard deviation of the error instead of the sketch size. 

Nice. The sketch setup is algorithm-specific thus normal users may easily get lost there.